librz/core/rop.c

```diff
@@ -0,0 +1,1248 @@
+	ut8 *buf;
+	RzIOMap *map;
+	HtSU *unique_hitlists = ht_su_new(HT_STR_DUP);
+
+	const ut64 search_from = rz_config_get_i(core->config, "search.from"),
+		   search_to = rz_config_get_i(core->config, "search.to");
+	if (search_from > search_to && search_to) {
+		RZ_LOG_ERROR("core: search.from > search.to is not supported\n");
+		ret = false;
+		goto bad;
+	}
+	// {.addr = UT64_MAX, .size = 0} means search range is unspecified
+	RzInterval search_itv = { search_from, search_to - search_from };
+	bool empty_search_itv = search_from == search_to && search_from != UT64_MAX;
+	if (empty_search_itv) {
+		RZ_LOG_ERROR("core: `from` address is equal `to`\n");
+		ret = false;
+		goto bad;
+	}
+	// TODO full address cannot be represented, shrink 1 byte to [0, UT64_MAX)
+	if (search_from == UT64_MAX && search_to == UT64_MAX) {
+		search_itv.addr = 0;
+		search_itv.size = UT64_MAX;
+	}
+
+	Sdb *gadgetSdb = NULL;
+	if (rz_config_get_i(core->config, "rop.sdb")) {
+		if (!(gadgetSdb = sdb_ns(core->sdb, "gadget_sdb", false))) {
+			gadgetSdb = sdb_ns(core->sdb, "gadget_sdb", true);
+		}
+	}
+	if (max_count == 0) {
+		max_count = -1;
+	}
+	if (max_instr <= 1) {
+		rz_list_free(end_list);
+		RZ_LOG_ERROR("core: ROP length (rop.len) must be greater than 1.\n");
+		if (max_instr == 1) {
+			RZ_LOG_ERROR("core: For rop.len = 1, use /c to search for single "
+				     "instructions. See /c? for help.\n");
+		}
+		return false;
+	}
+
+	if (!strcmp(arch, "mips")) { // MIPS has no jump-in-the-middle
+		increment = 4;
+	} else if (!strcmp(arch, "arm")) { // ARM has no jump-in-the-middle
+		increment = rz_config_get_i(core->config, "asm.bits") == 16 ? 2 : 4;
+	} else if (!strcmp(arch, "avr")) { // AVR is halfword aligned.
+		increment = 2;
+	}
+
+	if (greparg) {
+		grep_arg = strdup(greparg);
+		grep_arg = rz_str_replace(grep_arg, ",,", ";", true);
+	}
+
+	// Deal with the grep guy.
+	if (grep_arg && regexp) {
+		if (!rx_list) {
+			rx_list = rz_list_newf(free);
+		}
+		gregexp = strdup(grep_arg);
+		tok = strtok(gregexp, ";");
+		while (tok) {
+			rx = strdup(tok);
+			rz_list_append(rx_list, rx);
+			tok = strtok(NULL, ";");
+		}
+	}
+	rz_cmd_state_output_array_start(state);
+	rz_cons_break_push(NULL, NULL);
+	const char *mode_str = rz_config_get(core->config, "search.in");
+	boundaries = rz_core_get_boundaries_prot(core, -1, mode_str, "search");
+	if (!boundaries) {
+		rz_cmd_state_output_array_end(state);
+	}
+	rz_list_foreach (boundaries, itermap, map) {
+		HtUUOptions opt = { 0 };
+		HtUU *badstart = ht_uu_new_opt(&opt);
+		if (!rz_itv_overlap(search_itv, map->itv)) {
+			continue;
+		}
+		RzInterval itv = rz_itv_intersect(search_itv, map->itv);
+		ut64 from = itv.addr, to = rz_itv_end(itv);
+		if (rz_cons_is_breaked()) {
+			break;
+		}
+		delta = to - from;
+		buf = calloc(1, delta);
+		if (!buf) {
+			result = false;
+			goto bad;
+		}
+		(void)rz_io_read_at(core->io, from, buf, delta);
+
+		// Find the end gadgets.
+		for (i = 0; i < delta; i += increment) {
```

Comment:
```suggestion
		for (size_t i = 0; i < delta; i += increment) {
```

No need to define it above. Always has the risk that it is not reset properly.

Reply:
Can this whole loop be move to it's own `static` function? It would make the code way more readable.